File: 002_reverse_words/ReverseWords.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

vector<string>& split(const string& s, char delim, vector<string>& elems) {
    stringstream ss(s);
    string item;
    while (getline(ss, item, delim)) {
        elems.push_back(item);
    }

    return elems;
}

vector<string> split(const string& s, char delim) {
    vector<string> elems;
    split(s, delim, elems);
    return elems;
}

string solution1(string input) {
    vector<string> items = split(input, ' ');
    reverse(items.begin(), items.end());
    string output = "";
    for (unsigned int i = 0; i < items.size(); ++i) {
        if (i < items.size() - 1) {
            output += (items[i] + " ");
        } else {
            output += items[i];
        }
    }

    return output;
}

int main() {
    string testString = "How now brown cow";
    cout << solution1(testString) << endl;

    return 0;
}

File: 002_reverse_words/ReverseWords.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;

class ReverseWords {

    private static String solution1(String input) {
        ArrayList<String> items 
            = new ArrayList<String>(Arrays.asList(input.split(" ")));
        Collections.reverse(items);
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < items.size(); ++i) {
            if (i < items.size() - 1) {
                out.append(items.get(i)).append(" ");
            } else {
                out.append(items.get(i));
            }
        }

        return out.toString();
    }

    public static void main(String[] args) {
        String testString = "How now brown cow";
        System.out.println(solution1(testString));
    }
}
